src/api/cmd/chart-repo-sync/sync_test.go

```diff
@@ -0,0 +1,447 @@
+	tests := []struct {
+		name     string
+		repoYAML string
+	}{
+		{"invalid", "invalid"},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			_, err := parseRepoIndex([]byte(tt.repoYAML))
+			assert.ExistsErr(t, err, tt.name)
+		})
+	}
+
+	t.Run("valid", func(t *testing.T) {
+		index, err := parseRepoIndex([]byte(validRepoIndexYAML))
+		assert.NoErr(t, err)
+		assert.Equal(t, len(index.Entries), 2, "number of charts")
+		assert.Equal(t, index.Entries["acs-engine-autoscaler"][0].GetName(), "acs-engine-autoscaler", "chart version populated")
+	})
+}
+
+func Test_chartsFromIndex(t *testing.T) {
+	r := repo{Name: "test", URL: "http://testrepo.com"}
+	index, _ := parseRepoIndex([]byte(validRepoIndexYAML))
+	charts := chartsFromIndex(index, r)
+	assert.Equal(t, len(charts), 2, "number of charts")
+
+	indexWithDeprecated := validRepoIndexYAML + `
+  deprecated-chart:
+  - name: deprecated-chart
+    deprecated: true`
+	index2, err := parseRepoIndex([]byte(indexWithDeprecated))
+	assert.NoErr(t, err)
+	charts = chartsFromIndex(index2, r)
+	assert.Equal(t, len(charts), 2, "number of charts")
+}
+
+func Test_newChart(t *testing.T) {
+	r := repo{Name: "test", URL: "http://testrepo.com"}
+	index, _ := parseRepoIndex([]byte(validRepoIndexYAML))
+	c := newChart(index.Entries["wordpress"], r)
+	assert.Equal(t, c.Name, "wordpress", "correctly built")
+	assert.Equal(t, len(c.ChartVersions), 2, "correctly built")
+	assert.Equal(t, c.Description, "new description!", "takes chart fields from latest entry")
+	assert.Equal(t, c.Repo, r, "repo set")
+	assert.Equal(t, c.ID, "test/wordpress", "id set")
+}
+
+func Test_importCharts(t *testing.T) {
+	m := &mock.Mock{}
+	// Ensure Upsert func is called with some arguments
+	m.On("Upsert", mock.Anything)
+	m.On("RemoveAll", mock.Anything)
+	dbSession = mockstore.NewMockSession(m)
+	index, _ := parseRepoIndex([]byte(validRepoIndexYAML))
+	charts := chartsFromIndex(index, repo{Name: "test", URL: "http://testrepo.com"})
+	importCharts(charts)
+
+	m.AssertExpectations(t)
+	// The Bulk Upsert method takes an array that consists of a selector followed by an interface to upsert.
+	// So for x charts to upsert, there should be x*2 elements (each chart has it's own selector)
+	// e.g. [selector1, chart1, selector2, chart2, ...]
+	args := m.Calls[0].Arguments.Get(0).([]interface{})
+	assert.Equal(t, len(args), len(charts)*2, "number of selector, chart pairs to upsert")
+	for i := 0; i < len(args); i += 2 {
+		c := args[i+1].(chart)
+		assert.Equal(t, args[i], bson.M{"_id": "test/" + c.Name}, "selector")
+	}
+}
+
+func Test_fetchAndImportIcon(t *testing.T) {
+	t.Run("no icon", func(t *testing.T) {
+		c := chart{ID: "test/acs-engine-autoscaler"}
+		assert.NoErr(t, fetchAndImportIcon(c))
+	})
+
+	index, _ := parseRepoIndex([]byte(validRepoIndexYAML))
+	charts := chartsFromIndex(index, repo{Name: "test", URL: "http://testrepo.com"})
+
+	t.Run("failed download", func(t *testing.T) {
```

Comment:
Very cool use the subtest cases (t.Run) I was not aware of it :)

Reply:
new feature in 1.7, I really like it!